Src/FluentAssertions/Specialized/TaskCompletionSourceAssertions.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions.Common;
+using FluentAssertions.Execution;
+
+namespace FluentAssertions.Specialized
+{
+    public class TaskCompletionSourceAssertions<T>
+    {
+        private readonly TaskCompletionSource<T> subject;
+
+        public TaskCompletionSourceAssertions(TaskCompletionSource<T> tcs)
+            : this(tcs, new Clock())
+        {
+        }
+
+        public TaskCompletionSourceAssertions(TaskCompletionSource<T> tcs, IClock clock)
+        {
+            subject = tcs;
+            Clock = clock ?? throw new ArgumentNullException(nameof(clock));
+        }
+
+        private protected IClock Clock { get; }
+
+        /// <summary>
+        /// Asserts that the <see cref="Task"/> of the current <see cref="TaskCompletionSource{T}"/> will complete within the specified time.
+        /// </summary>
+        /// <param name="timeSpan">The allowed time span for the operation.</param>
+        /// <param name="because">
+        /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+        /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <see cref="because" />.
+        /// </param>
+        public Task<AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>> CompleteWithinAsync(
+            TimeSpan timeSpan, string because = "", params object[] becauseArgs)
+        {
+            return AssertAsync(timeSpan, because, becauseArgs,
+                "Expected {context} to complete within {0}{reason}, but found <null>.",
+                "Expected {context:task} to complete within {0}{reason}.",
+                new object[] { timeSpan });
+        }
+
+        private async Task<AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>> AssertAsync(
```

Comment:
You *could* inline this method again to avoid parameter names such as `failMessage1` and `failMessage2`. 
Not crucial since it's not part of the public API.

Reply:
I think, this is a good entry for future extensions.